scripts/custom-linter.js

```diff
@@ -0,0 +1,167 @@
+
+  for (const inputFilePath of inputFilePaths) {
+    const inputCode = fs.readFileSync(inputFilePath, 'utf8');
+    lintOneFile(inputFilePath, inputCode);
+  }
+
+  const violationCountAnsi = colors.bold(violationCount);
+  const plural = violationCount === 1 ? '' : 's';
+  console.log(`\n${violationCountAnsi} Custom Linter error${plural}\n`);
+}
+
+/**
+ * @param {string} inputFilePath
+ * @param {string} inputCode
+ */
+function lintOneFile(inputFilePath, inputCode) {
+  checkAllImportPaths(inputFilePath, inputCode);
+}
+
+/**
+ * @param {string} inputFilePath
+ * @param {string} inputCode
+ */
+function checkAllImportPaths(inputFilePath, inputCode) {
+  const ast = getAstFromCodeString(inputCode);
+
+  traverse(ast, {
+    ExportDeclaration(oldExportDeclaration) {
+      checkOneImportPath(oldExportDeclaration, inputFilePath);
+    },
+    ImportDeclaration(oldImportDeclaration) {
+      checkOneImportPath(oldImportDeclaration, inputFilePath);
+    },
+  });
+}
+
+/**
+ * @param {!Object} importOrExportDeclaration
+ * @param {string} inputFilePath
+ */
+function checkOneImportPath(importOrExportDeclaration, inputFilePath) {
+  // Ensure exports have a `from '...'` qualifier.
+  if (!importOrExportDeclaration.node.source) {
+    return;
+  }
+
+  const inputFileDir = path.dirname(inputFilePath);
+
+  /** @type {string} */
+  const importPath = importOrExportDeclaration.node.source.value;
+  const importPathAnsi = colors.bold(importPath);
+
+  if (new RegExp('^@material/[^/]+$').test(importPath)) {
+    logLinterViolation(
+      inputFilePath,
+      // eslint-disable-next-line max-len
+      `Import path '${importPathAnsi}' resolves to a transpiled JS file instead of a TS source file. Please include the extension-less filename of the specific file you wish to import. E.g.: '${importPath}/index'`
+    );
+  }
+
+  if (importPath.indexOf('../mdc-') > -1 || importPath.indexOf('../material-components-web') > -1) {
+    logLinterViolation(
+      inputFilePath,
+      // eslint-disable-next-line max-len
+      `Import path '${importPathAnsi}' appears to be a relative path to another package. Please use a @material import instead. E.g.: import '@material/foo/types' instead of import '../mdc-foo/types'.`
+    );
+  }
+
+  if (importPath.indexOf('./') > -1) {
+    /** @type {string} */
+    const resolvedPath = resolve.sync(importPath, {basedir: inputFileDir, extensions: ['.ts', '.js']});
+    let relativePath = path.relative(inputFileDir, resolvedPath);
+    if (!relativePath.startsWith('./') && !relativePath.startsWith('../')) {
+      relativePath = `./${relativePath}`;
+    }
+    const pointsToTSFile = relativePath === `${importPath}.ts`;
+    if (!pointsToTSFile) {
+      logLinterViolation(
+        inputFilePath,
+        // eslint-disable-next-line max-len
+        `Import path '${importPathAnsi}' should point directly to a specific TypeScript file. E.g.: import '${relativePath.replace('.ts', '')}'`
```

Comment:
I'm confused... doesn't this block get executed when `relativePath` _doesn't_ have `.ts`? So what does replacing `.ts` with nothing accomplish?

Reply:
`resolvedPath` is the full, absolute path to the TS file, including `.ts` extension.

`relativePath` is the relative version of `resolvedPath`, so it still has a `.ts` extension.

I renamed the variables to make this clearer. Sorry for the confusion!

Reply:
My point was this statement is within an `if (!pointsToTSFile)` block, which I had interpreted (based on the previous line of code before the condition) as meaning that the filename doesn't end with `.ts`, which means there's no `'.ts'` to replace with `''` anyway. Did I misinterpret the line before this condition?

Reply:
Gotcha. I refactored the code and was able to remove this entire `if` block. A different `if` statement now handles the same use case in a more general way.